### ISSUE TITLE
Top-K Results Not Working Despite January Confirmation - Classification Model

### ISSUE THREAD

**Issue by curtinmjc** [ultralytics/hub#1141]
### Search before asking

- [x] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and found no similar bug report.

### HUB Component

Inference

### Bug

Following up on January 23rd comment by glenn-jocher in issue #906 where he stated:

> "For classification models... the top-k results are retuned, i.e. top-1 or top-5"

I'm still only receiving single top predictions from my classification model via HUB API, not the top-K results described.

**Current API Response (only top-1):**
```json
{
  "images": [{"results": [{"class": "class1", "confidence": 0.85}]}]
}
```

Expected (based on glenn-jocher comment):  Top-5 results with confidences for classification models

Questions:
1. Is there a specific parameter needed to get top-5 results?  (I tried adding `top_k=5` parameter to API request and it did not work)
2. Is there a bug preventing top-K from being returned?

Model Type: YOLOv8 Classification

**Comment by UltralyticsAssistant** [comment]
👋 Hello @curtinmjc, thank you for raising an issue about [Ultralytics HUB](https://hub.ultralytics.com) 🚀! Please visit our [HUB Docs](https://docs.ultralytics.com/hub) to learn more:
- [**Quickstart**](https://docs.ultralytics.com/hub/quickstart). Start training and deploying YOLO models with HUB in seconds.
- [**Datasets: Preparing and Uploading**](https://docs.ultralytics.com/hub/datasets). Learn how to prepare and upload your datasets to HUB in YOLO format.
- [**Projects: Creating and Managing**](https://docs.ultralytics.com/hub/projects). Group your models into projects for improved organization.
- [**Models: Training and Exporting**](https://docs.ultralytics.com/hub/models). Train YOLOv5 and YOLOv8 models on your custom datasets and export them to various formats for deployment.
- [**Integrations**](https://docs.ultralytics.com/hub/integrations). Explore different integration options for your trained models, such as TensorFlow, ONNX, OpenVINO, CoreML, and PaddlePaddle.
- [**Ultralytics HUB App**](https://docs.ultralytics.com/hub/app). Learn about the Ultralytics App for iOS and Android, which allows you to run models directly on your mobile device.
    * [**iOS**](https://docs.ultralytics.com/hub/app/ios). Learn about YOLO CoreML models accelerated on Apple's Neural Engine on iPhones and iPads.
    * [**Android**](https://docs.ultralytics.com/hub/app/android). Explore TFLite acceleration on mobile devices.
- [**Inference API**](https://docs.ultralytics.com/hub/inference-api). Understand how to use the Inference API for running your trained models in the cloud to generate predictions.

If this is a 🐛 Bug Report, please provide screenshots and steps to [reproduce](https://docs.ultralytics.com/help/minimum_reproducible_example/) your problem to help us get started working on a fix. A minimum reproducible example (MRE) is especially helpful to pinpoint the issue.

If this is a ❓ Question, please provide as much information as possible, including dataset, model, environment details etc. so that we might provide the most helpful response.

This is an automated response 🤖. An Ultralytics engineer will review your issue and assist you soon. Thank you for your patience!

**Comment by curtinmjc** [comment]
Thank you for the quick response.  Below if the additional information you requested:

**API Request Format:**
- Endpoint: POST https://api.ultralytics.com/v1/predict/{model_id}
- Headers: x-api-key  
- Form data: file (image), imgsz=448
- No additional parameters

**Model:** YOLOv8l classification model with 4926 classes

**Note:** Can provide actual model ID and sample request privately if needed for debugging.

**Comment by curtinmjc** [comment]
Do you have an update for me?

**Comment by pderrenger** [comment]
Hi @curtinmjc! 

Thank you for your patience and for providing the additional details. I appreciate the comprehensive information about your API request format and model specifications.

## Current Status

Based on your information and investigation into the current [Ultralytics HUB Inference API](https://docs.ultralytics.com/hub/inference-api/), this appears to be a **confirmed limitation** where the HUB API is not currently exposing top-K functionality for classification models, despite the underlying YOLO classification models having this capability.

Your API request format looks correct:
```bash
POST https://api.ultralytics.com/v1/predict/{model_id}
Headers: x-api-key
Form data: file, imgsz=448
```

## Workaround Options

While we work on addressing this in the HUB API, you could consider these alternatives:

1. **Local Inference**: Use the Ultralytics Python package directly for top-K results:
```python
from ultralytics import YOLO

model = YOLO('path/to/your/model.pt')
results = model('image.jpg')

# Access top-5 predictions
probs = results[0].probs
top5_indices = probs.top5
top5_confidences = probs.top5conf
```

2. **Export and Deploy**: Export your model and deploy it with a custom inference service that supports top-K results.

## Next Steps

This issue has been logged for the engineering team to investigate implementing top-K parameter support in the HUB Inference API. The discrepancy between the January confirmation and current API behavior needs to be resolved.

We'll keep this issue open to track progress on adding proper top-K support to the classification inference API. Thank you for your detailed reporting - it helps us improve the platform! 🚀

**Comment by curtinmjc** [comment]
As a result of the Top-K results problem of this Issue #1141 ("confirmed limitation where the HUB API is not currently exposing top-K functionality for classification models, despite the underlying YOLO classification models having this capability"), I followed your recommendation and created a "custom inference service that supports top-K results" on my own server.

But yesterday I ran some old code that hits an Ultralytics HUB YOLO classification model and was surprised to see the JSON classification response now includes 5 predictions, not just 1.

So does this mean you have updated the Ultralytics HUB Inference API and this issue is resolved?

**Comment by curtinmjc** [comment]
Thank you for the quick response.  Responses to your requests:
1. Classification request:  My app is sending a POST with multipart/form-data.  Below is an example:

Ultralytics Request Details:
   URL: https://predict.ultralytics.com
   Fields: ["imgsz": "640", "conf": "0.8", "iou": "0.45", "model": "https://hub.ultralytics.com/models/ULBOh8bbUFodt3dznuKl"]

2. Sample of YOLO shared model response is below:

{
  "images" : [
    {
      "results" : [
        {
          "class" : 4314,
          "name" : "Name of 4314",
          "confidence" : 0.99414999999999998
        },
        {
          "class" : 750,
          "name" : "Name of 750",
          "confidence" : 0.0042599999999999999
        },
        {
          "class" : 1931,
          "name" : "Name of 1931",
          "confidence" : 0.00038000000000000002
        },
        {
          "class" : 2587,
          "name" : "Name of 2587",
          "confidence" : 0.00032000000000000003
        },
        {
          "class" : 4641,
          "name" : "Name of 4641",
          "confidence" : 0.00022000000000000001
        }
      ],
      "shape" : [
        1512,
        2016
      ],
      "speed" : {
        "postprocess" : 0.057299999999999997,
        "preprocess" : 39.354170000000003,
        "inference" : 1418.0315000000001
      }
    }
  ],
  "metadata" : {
    "functionTimeCall" : 3.4180000000000001,
    "functionTimeAlive" : 64450.498,
    "model" : "https:\/\/hub.ultralytics.com\/models\/ULBOh8bbUFodt3dznuKl",
    "imageCount" : 1,
    "version" : {
      "torchvision" : "0.24.1+cpu",
      "python" : "3.11.14",
      "torch" : "2.9.1+cpu",
      "ultralytics" : "8.4.6"
    }
  }
}

**Comment by pderrenger** [comment]
Yes — based on the request/response you pasted, this is now behaving as expected for classification: the **Shared Inference** gateway at `https://predict.ultralytics.com` is returning a **top‑5 list** under `images[0].results` (each item includes `class`, `name`, `confidence`). That means the original “only top‑1 is returned” problem from #1141 is effectively **resolved from the API consumer perspective**.

A couple important notes from your example:

### 1) There’s still no `top_k` request parameter (top‑5 is the default)
Today the shared response is a fixed **top‑5** list for classification. Under the hood, Ultralytics classification results expose this through the `Probs` container (`top5` / `top5conf`). See the [`Probs` reference](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Probs) for the exact properties.

Related: the Python-side `Results.summary()` implementation for classification iterates `probs.top5` / `probs.top5conf` (i.e., top‑5 by default). See the [`Results.summary()` reference](https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.summary).

If you want “top‑3” (or “top‑1”) you can just take the first N entries client-side:
- top‑1: `results[0]`
- top‑3: `results[:3]`

### 2) Your `conf=0.8` isn’t filtering classification entries
In your response, entries #2–#5 are well below 0.8, which strongly suggests the **`conf` field is not applied as a cutoff for classification** in this endpoint (it’s primarily meaningful for detection-type outputs). If you want a minimum confidence threshold for classification, you’ll want to filter after you receive the response, e.g. keep only items where `confidence >= 0.8`.

### 3) Your metadata confirms you’re on the newer behavior
Your response metadata shows `ultralytics: 8.4.6`, and that aligns with the top‑5 behavior you’re seeing now (server-side runtime).

---

If you’re happy with “top‑5 always” then we can consider #1141 closed/resolved ✅. If you need a true configurable `top_k` (top‑10, top‑20, etc.), that would be a separate enhancement request — tell me the `k` you need and whether you’d want it returned as additional items in `images[].results` or under a separate key.

**Comment by pderrenger** [comment]
Totally fair feedback, and I agree: **the underlying Ultralytics YOLO classification stack can produce an arbitrary top‑K**, so having the **Shared Inference** response locked to “top‑5 always” is limiting—especially for large‑label models like your 4,926‑class classifier.

### Where things stand today
- `https://predict.ultralytics.com` currently returns a **fixed top‑5** list for classification (as you’re seeing now), and there’s **no supported `top_k` request field** at the moment.
- This matches the core library’s “top‑5 by default” behavior (i.e., `Probs.top5` / `Probs.top5conf`, and `Results.summary()` iterating over those). See:
```text
Probs reference: https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Probs
Results.summary() reference: https://docs.ultralytics.com/reference/engine/results/#ultralytics.engine.results.Results.summary
```

### Why it may have been fixed to top‑5 (context)
Top‑5 is a common convention for classification and keeps **response size + latency** predictable. But agreed: **it shouldn’t prevent power users from requesting top‑10/top‑20**, etc.

### Best next step (to get this implemented)
Since #1141 was about “top‑1 only” (now resolved), the cleanest way to track this is a **new feature request** in `ultralytics/hub` referencing #1141.

If you do that, can you include:
1. **What `k` do you need** (e.g., 10? 20? 100?) and whether you need a max cap.
2. Desired behavior when `top_k` is omitted (keep default top‑5?).
3. Whether you’d want **names included** (payload size impact) or just `class` + `confidence`.

If you reply here with your preferred `k` + schema, I’ll mirror it into an internal request on our side as well. 👍

### Workaround recap (until `top_k` exists server-side)
- If you only need ≤5: you’re set already (slice client-side for top‑1/top‑3).
- If you need >5: your custom service approach is still the right workaround, because the shared endpoint isn’t returning the full probability vector.

For completeness, here’s how you can compute arbitrary top‑K locally from the full probs tensor:
```python
from ultralytics import YOLO
import torch

model = YOLO("your_model.pt")
r = model("image.jpg")[0]
p = r.probs  # Probs
k = 20

# Get top-k from full distribution:
vals, idx = torch.topk(p.data, k)  # p.data is the raw probs tensor
topk = [{"class": int(i), "name": r.names[int(i)], "confidence": float(v)} for v, i in zip(vals, idx)]
```

If you tell me the exact `k` you’re targeting (and whether you need “names” in the response), we can propose a concrete `top_k` contract that won’t surprise existing users.
